src/app/modules/shared_modules/keycard_popup/module.nim

```diff
@@ -228,43 +257,86 @@ proc prepareKeyPairsModel[T](self: Module[T]) =
+          keyUid = a.keyUid,
+          locked = false,
           name = "Key " & $(diffImports + 1), # string created here should be transalted, but so far it's like it is
           image = "",
           icon = "key_pair_private_key",
-          pairType = KeyPairType.SeedImport,
+          pairType = KeyPairType.PrivateKeyImport,
           derivedFrom = a.derivedfrom)
         items.add(item)
       item.addAccount(a.name, a.path, a.address, a.emoji, a.color, icon = "", balance = 0.0)
       continue
   if items.len == 0:
     debug "sm_there is no any key pair for the logged in user that is not already migrated to a keycard"
-  self.view.createKeyPairModel(items)
-
-method runFlow*[T](self: Module[T], flowToRun: FlowType) =
+  return items
+
+proc prepareKeyPairItemForAuthentication[T](self: Module[T], keyUid: string) =
+  var item = initKeyPairItem()
+  self.view.createKeyPairForAuthentication()
+  let items = self.buildKeyPairsList()
+  for it in items:
+    if it.keyUid == keyUid:
+      item = it
+      break
+  if item.name.len == 0:
+    error "sm_cannot find keypair among known keypairs for the given keyUid", keyUid=keyUid
+  item.setPubKey("")
+  item.setImage("")
+  item.setIcon("keycard")
+  item.setPairType(KeyPairType.Unknown)
+  self.view.setKeyPairForAuthentication(item)
+
+method runFlow*[T](self: Module[T], flowToRun: FlowType, keyUid = "", bip44Path = "", txHash = "") =
+  ## In case of `Authentication` if we're signing a transaction we need to provide a key uid of a keypair that an account 
+  ## we want to sign a transaction for belongs to. If we're just doing an authentication for a logged in user, then 
+  ## default key uid is always the key uid of the logged in user.
   if flowToRun == FlowType.General:
     self.controller.terminateCurrentFlow(lastStepInTheCurrentFlow = false)
     error "sm_cannot run an general flow"
     return
   if not self.initialized:
+    self.initialized = true
     self.controller.init()
   if flowToRun == FlowType.FactoryReset:
-    self.prepareKeyPairsModel()
+    let items = self.buildKeyPairsList()
+    self.view.createKeyPairModel(items)
     self.tmpLocalState = newReadingKeycardState(flowToRun, nil)
     self.controller.runGetMetadataFlow()
     return
   if flowToRun == FlowType.SetupNewKeycard:
-    self.prepareKeyPairsModel()
+    let items = self.buildKeyPairsList()
+    self.view.createKeyPairModel(items)
     self.view.setCurrentState(newSelectExistingKeyPairState(flowToRun, nil))
     self.controller.readyToDisplayPopup()
     return
+  if flowToRun == FlowType.Authentication:
+    self.controller.connectKeychainSignals()
+    if keyUid.len > 0:
+      self.prepareKeyPairItemForAuthentication(keyUid)
+      self.tmpLocalState = newReadingKeycardState(flowToRun, nil)
+      self.controller.runSignFlow(keyUid, bip44Path, txHash)
+      return
+    if self.controller.loggedInUserUsesBiometricLogin():
+      self.controller.tryToObtainDataFromKeychain()
+      return
```

Comment:
Wondering what happens here? What are the expected signals/events that ensure we don't get into a locked state?
Looking into `tryToObtainDataFromKeychain` if `loggedInUserUsesBiometricLogin` is false then we return and don't follow up. Is there a backup mechanism I'm missing here?

Reply:
@stefandunca you missed it a bit here :) we're returning if `loggedInUserUsesBiometricLogin` is true. So in that case we just call `tryToObtainDataFromKeychain` and return and wait for the signal coming from the keychain service in order to decide what to do next.

Reply:
Now I see it is the same check. Thanks for the clarifications.